addons/website_sale/controllers/product_configurator.py

```diff
@@ -1,147 +1,263 @@
+
+from odoo.addons.sale.controllers.product_configurator import SaleProductConfiguratorController
 from odoo.addons.website_sale.controllers.main import WebsiteSale
 
-class WebsiteSaleProductConfiguratorController(WebsiteSale):
+
+class WebsiteSaleProductConfiguratorController(SaleProductConfiguratorController, WebsiteSale):
 
     @route(
-        '/sale_product_configurator/show_advanced_configurator',
-        type='json', auth='public', methods=['POST'], website=True,
+        route='/website_sale_product_configurator/show_dialog',
+        type='json',
+        auth='public',
+        website=True,
     )
-    def show_advanced_configurator(
-        self, product_id, variant_values, add_qty=1, force_dialog=False, **kw,
+    def website_sale_product_configurator_show_dialog(
+        self, product_template_id, ptav_ids, is_product_configured
     ):
-        product = request.env['product.product'].browse(int(product_id))
-        product_template = product.product_tmpl_id
-        combination = request.env['product.template.attribute.value'].browse(variant_values)
-        has_optional_products = product.optional_product_ids.filtered(
-            lambda p: p._is_add_to_cart_possible(combination)
-                      and (not request.website.prevent_zero_price_sale or p._get_contextual_price())
-        )
+        """ Return whether the product configurator dialog should be shown.
 
-        already_configured = bool(combination)
-        if not force_dialog and not has_optional_products and (
-            product.product_variant_count <= 1 or already_configured
-        ):
-            # The modal is not shown if there are no optional products and
-            # the main product either has no variants or is already configured
-            return False
-
-        add_qty = float(add_qty)
-        combination_info = product_template._get_combination_info(
-            combination=combination,
-            product_id=product.id,
-            add_qty=add_qty,
+        :param int product_template_id: The product being checked, as a `product.template` id.
+        :param list(int) ptav_ids: The combination of the product, as a list of
+                                   `product.template.attribute.value` ids.
+        :param bool is_product_configured: Whether the product is already configured.
+        :rtype: bool
+        :return: Whether the product configurator dialog should be shown.
+        """
+        product_template = request.env['product.template'].browse(product_template_id)
+        combination = request.env['product.template.attribute.value'].browse(ptav_ids)
+        single_product_variant = product_template.get_single_product_variant()
+        # We can't use `single_product_variant.get('has_optional_products')` as it doesn't take
+        # `combination` into account.
+        has_optional_products = bool(product_template.optional_product_ids.filtered(
+            lambda op: self._show_product(op, combination)
+        ))
```

Comment:
Hmm, I'd need to investigate this. On one hand, I agree it would be a bit cleaner, but on the other hand, we'd need to do exactly the same in `_should_show_product` (which filters out optional products that shouldn't be shown in the configurator) and `get_single_product_variant` (which indicates whether a given product has optional products). Otherwise there may be inconsistencies in the configurator. But I'm not entirely sure these 2 methods should return the same results.